### ISSUE TITLE
Improve nullable reference type handling with NUnit [SetUp]

### ISSUE THREAD

**Issue by sbergen** [SonarSource/sonar-dotnet#9522]
## Issue description

With nullable reference types enabled, a `CS8618 Non-nullable field must contain a non-null value when exiting constructor` warning is (expectedly) emitted with the following code (using NUnit):
```
class MyTest
{
	private object field;

	[SetUp]
	public void SetUp()
	{
		this.field = new();
	}
}
```

This warning is suppressed by some other tools (namely Rider), as it's not really useful. It would be nice if Sonar would also suppress it.

## Context

While this is not really a false positive, e.g. Rider contains logic to suppress the warning, when the field is assigned in a method with the `SetUp` or `OneTimeSetUp` attribute, and having a disparity between IDE and Sonar behaviour is somewhat problematic.

Also, I am aware that using `FixtureLifeCycleAttribute` with `LifeCycle.InstancePerTestCase` works much nicer with NRT enabled, but since we are working in Unity, which is stuck with an ancient version of NUnit, we can not use this.

**Comment by martin-strecker-sonarsource** [comment]
Hello @sbergen,

CS8618 is a warning from the csharp compiler, and we don't have a mechanism to suppress these warnings. Microsoft describes a couple of possible solutions to get rid of the warning in [this article](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings#nonnullable-reference-not-initialized). The canonical way in a case like yours is to use a field initializer with a [null-forgiving operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving):
```cs
class MyTest
{
	private object field = default!;

	[SetUp]
	public void SetUp()
	{
		this.field = new();
	}
}
``` 
This pattern is also used for [entity classes](https://learn.microsoft.com/en-us/ef/core/miscellaneous/nullable-reference-types#non-nullable-properties-and-initialization) which face the same problem.

**Comment by sbergen** [comment]
Thanks for detailed answer! However, how to get rid of the warning was not really my issue:

> having a disparity between IDE and Sonar behaviour is somewhat problematic

The issue I was trying to outline is delayed feedback in the environment I work in: Rider suppresses this warning in this specific context, so I'm only guaranteed to see the warning once my changes hit CI and get inspected by Sonar. (The warning might pop up in Unity also, but it's easy to miss, as it may get cleared on a following compilation.)

If this is still a "won't do", I won't push any further - I know I'm working in a very specific environment. Just wanted to make sure we are on the same page here.

**Comment by martin-strecker-sonarsource** [comment]
The [Roslyn analyzer APIs](https://learn.microsoft.com/en-us/visualstudio/extensibility/getting-started-with-roslyn-analyzers?view=vs-2022) do not support suppressing issues raised by other analyzers (in your case, Roslyn itself). Tools like Rider or our IDE plugin SonarLint can filter out such warnings on an IDE-to-IDE basis.
Technically, there is nothing we can do here because the issue flow is like this:
* In the IDE: Rider IDE -> runs the Roslyn compiler in the background to gather compiler warnings -> Rider filters the reported issues
* In CI: The Roslyn compiler generates compiler warnings as part of the compilation -> The reported issues are imported into SonarCloud or SonarQube.

The analyzer developed in this Repo is run along with the Roslyn compilation and can raise additional issues. It can not suppress any issues reported by another [DiagnosticAnalyzer](https://github.com/dotnet/roslyn/blob/main/docs/wiki/How-To-Write-a-C%23-Analyzer-and-Code-Fix.md), like the one that raises CS8618. Therefore, it is reported to the server, and we do not have the logic to filter out any issues in the report.

I see how this is cumbersome for you, but I do not see any option for fixing it.

**Comment by killergege** [comment]
We have the same question with Visual Studio.

One thing I don't understand is that we use the [nunit.analyzers](https://github.com/nunit/nunit.analyzers) nuget package, which, according to its documentation, is able to suppress the warning ([this rule](https://docs.nunit.org/articles/nunit-analyzers/NUnit3002.html) for that specific case). Which seem to contradict what you are saying (that the API does not support suppressing issues).
Though it seems to work with VS, as this warning does not appear in the IDE, but not in the CI.

**Comment by martin-strecker-sonarsource** [comment]
You are right. There is a way to suppress other diagnostics with a [DiagnosticSuppressor](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticsuppressor?view=roslyn-dotnet-3.3.1). That class was introduced with Roslyn 3.1.1. Unfortunately we are still using Roslyn 1.3 for compatibility reasons and can not use that API at the moment.
We will keep this in mind and come back to this issue when we move to a new Roslyn version.
